test/e2e/density.go

```diff
@@ -0,0 +1,178 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func RunRC(c *client.Client, test string, image string, replicas int) {
+	var last int
+	current := 0
+	same := 0
+
+	ns := api.NamespaceDefault
+	name := "my-hostname-" + test + "-" + string(util.NewUUID())
+
+	By(fmt.Sprintf("Creating replication controller %s", name))
+	rc, err := c.ReplicationControllers(ns).Create(&api.ReplicationController{
+		ObjectMeta: api.ObjectMeta{
+			Name: name,
+		},
+		Spec: api.ReplicationControllerSpec{
+			Replicas: replicas,
+			Selector: map[string]string{
+				"name": name,
+			},
+			Template: &api.PodTemplateSpec{
+				ObjectMeta: api.ObjectMeta{
+					Labels: map[string]string{"name": name},
+				},
+				Spec: api.PodSpec{
+					Containers: []api.Container{
+						{
+							Name:  name,
+							Image: image,
+							Ports: []api.Port{{ContainerPort: 80}},
+						},
+					},
+       				},
+			},
+		},
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	defer func () {
+		By("Cleaning up the replication controller")
+		rc.Spec.Replicas = 0
+		if _, err = c.ReplicationControllers(ns).Update(rc); err != nil {
+			Logf("Failed to resize replication controller %s to zero: %v", name, err)
+		}
+
+		// Delete the replication controller.
+		if err = c.ReplicationControllers(ns).Delete(name); err != nil {
+			Logf("Failed to delete replication controller %s: %v", name, err)
+		}
+	}()
+
+	By(fmt.Sprintf("Making sure all %d replicas exist", replicas))
+	label := labels.SelectorFromSet(labels.Set(map[string]string{"name": name}))
+	pods, err := c.Pods(ns).List(label)
+	Expect(err).NotTo(HaveOccurred())
+	fail_cnt := 5
+	for same < fail_cnt {
+		current = len(pods.Items)
+		Logf("Controller %s: Found %d pods out of %d", name, current, replicas)
+		if len(pods.Items) == replicas {
+			break
+		} else if last < current {
+			same = 0
+		} else if last == current {
+			same++
+		} else if current < last {
+			Failf("Controller %s: Number of submitted pods dropped from %d to %d", last, current)
+		}
+
+		if same >= fail_cnt {
+			Logf("No pods submitted for the last %d checks", fail_cnt)
+		}
+		
+		last = current
+		time.Sleep(5 * time.Second)
+		pods, err = c.Pods(ns).List(label)
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
Part of this test is not just that the system can handle x pods per node, but that it can do it without sacrificing other functionality.  We certainly wouldn't want the system to be running a large number of pods but be unresponsive to administrative functions like listing or removing pods.  If the system is running pods but it is too busy to respond to a simple query of the number of pods in the system then imo the system is too busy to be useful.